velox/vector/tests/MapVectorToRowVectorTest.cpp

```diff
@@ -0,0 +1,473 @@
+    testCases.push_back(map);
+
+    auto makeCopyWithValues = [&](VectorPtr& values) {
+      return std::make_shared<MapVector>(
+          pool(),
+          MAP(keyType, values->type()),
+          map->nulls(),
+          map->size(),
+          map->offsets(),
+          map->sizes(),
+          map->mapKeys(),
+          values);
+    };
+
+    // With String Values.
+    VectorFuzzer::Options opts{.nullRatio = 0};
+    VectorFuzzer fuzzer(opts, pool());
+    auto stringValues = fuzzer.fuzzFlat(VARCHAR(), intValues->size());
+    stringValues->setNulls(intValues->nulls());
+    testCases.push_back(makeCopyWithValues(stringValues));
+
+    // With Array Values.
+    auto arrayValues = fuzzer.fuzzFlat(ARRAY(BIGINT()), intValues->size());
+    arrayValues->setNulls(intValues->nulls());
+    testCases.push_back(makeCopyWithValues(arrayValues));
+
+    // With Row Values.
+    auto rowValues = fuzzer.fuzzFlat(ROW({BIGINT()}), intValues->size());
+    rowValues->setNulls(intValues->nulls());
+    testCases.push_back(makeCopyWithValues(rowValues));
+
+    // With Map Values.
+    auto mapValues =
+        fuzzer.fuzzFlat(MAP(BIGINT(), BIGINT()), intValues->size());
+    mapValues->setNulls(intValues->nulls());
+    testCases.push_back(makeCopyWithValues(mapValues));
+
+    auto opaqueValues =
+        BaseVector::create(OPAQUE<int>(), intValues->size(), pool());
+    auto flat = opaqueValues->asFlatVector<std::shared_ptr<void>>();
+    for (int i = 0; i < intValues->size(); i++) {
+      flat->set(i, std::make_shared<NonPOD>(i));
+    }
+    testCases.push_back(makeCopyWithValues(opaqueValues));
+
+    return testCases;
+  }
+
+  template <TypeKind ValueKind>
+  void verifyDefaultValue(VectorPtr& vector, vector_size_t row) {
+    if constexpr (
+        (TypeTraits<ValueKind>::isPrimitiveType ||
+         ValueKind == TypeKind::OPAQUE) &&
+        ValueKind != TypeKind::UNKNOWN) {
+      using NativeType = typename TypeTraits<ValueKind>::NativeType;
+      static NativeType defaultValue;
+      ASSERT_EQ(vector->asFlatVector<NativeType>()->valueAt(row), defaultValue);
+    } else if constexpr (
+        ValueKind == TypeKind::ARRAY || ValueKind == TypeKind::MAP) {
+      auto arrayBaseVector = vector->asChecked<ArrayVectorBase>();
+      ASSERT_NE(arrayBaseVector, nullptr);
+      ASSERT_EQ(arrayBaseVector->offsetAt(row), 0);
+      ASSERT_EQ(arrayBaseVector->sizeAt(row), 0);
+    } else {
+      VELOX_USER_FAIL(
+          "Unsupported type for replacing nulls: {}",
+          TypeTraits<ValueKind>::TypeTraits::name);
+    }
+  }
+
+  // Helper function that verifies the result of MapVector::toRowVector by
+  // iterating through all rows of the result and looking up its corresponding
+  // value in the original MapVector (by looking up the key for that row), if it
+  // exists. If key is not found, ensures that its set to null or the default
+  // value depending on whether replaceNulls is set.
+  void verifyResult(
```

Comment:
This code seems to duplicate the code being tested, which makes it hard to ensure it is correct. Wondering if there is a way to test in a more black-box manner. For example,

make_row_from_map(m, keys, names) is equivalent to cast(row_constructor(m[k1], m[k2], ...) as row(name1, name2,...))

Can we base the test on this invariant?